configs/example/gem5_library/x86-global-inst-tracker.py

```diff
@@ -0,0 +1,159 @@
+# documentation and/or other materials provided with the distribution;
+# neither the name of the copyright holders nor the names of its
+# contributors may be used to endorse or promote products derived from
+# this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+# A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+# OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+# LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+# DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+# THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+"""
+
+This script demonstrates how to use global and local instruction trackers to
+monitor and control the simulation based on the number of instructions
+committed.
+
+This script will create a global instruction tracker to manage each local
+instruction tracker and connects each local instruction tracker to a core.
+The global instruction tracker will raise an event when the number of
+instructions committed across all cores reaches a certain threshold.
+
+In this script, we expect to start monitorning the instruction committed from
+the start of the simulation and raise a SIMPOINT_BEGIN exit event when all
+cores in combination have committed 100,000,000 instructions. Then, we will
+change the threshold to 20,000 instructions and raise another SIMPOINT_BEGIN
+event when the new threshold is reached. Finally, we will stop listening to
+instructions.
+
+Usage:
+------
+
+scons build/X86/gem5.opt
+./build/X86/gem5.opt [--debug-flags=InstTracker] \
+    configs/example/gem5_library/x86-global-inst-tracker.py
+
+"""
+
+import m5
+from m5.objects import (
+    GlobalInstTracker,
+    LocalInstTracker,
+)
+
+from gem5.components.boards.simple_board import SimpleBoard
+from gem5.components.cachehierarchies.classic.private_l1_cache_hierarchy import (
+    PrivateL1CacheHierarchy,
+)
+from gem5.components.memory.single_channel import SingleChannelDDR4_2400
+from gem5.components.processors.cpu_types import CPUTypes
+from gem5.components.processors.simple_processor import SimpleProcessor
+from gem5.isas import ISA
+from gem5.resources.resource import obtain_resource
+from gem5.simulate.exit_event import ExitEvent
+from gem5.simulate.simulator import Simulator
+
+cache_hierarchy = PrivateL1CacheHierarchy(
+    l1d_size="64kB",
+    l1i_size="64kB",
+)
+
+memory = SingleChannelDDR4_2400("1GB")
+
+processor = SimpleProcessor(cpu_type=CPUTypes.TIMING, num_cores=9, isa=ISA.X86)
```

Comment:
You have to have (number of threads) + 1 cores in SE mode. So, if the workload has 8 threads, then you have to have 9 cores.

Reply:
If I record correctly, using openmp for SE mode needs a core to handle the scheduler, that's why it is using 9 cores instead of 8 cores here. Adding comments is a good idea. I'll add comments to explain this here.